irohad/main/impl/on_demand_ordering_init.cpp

```diff
@@ -0,0 +1,168 @@
+        } else {
+          current_reject_round_ = 1;
+
+          // retrieve peer list from database
+          peer_query_factory->createPeerQuery() | [](auto &&query) {
+            return query->getLedgerPeers();
+          } | [this](auto &&peers) { current_peers_ = std::move(peers); };
+
+          // generate permutation of peers list from corresponding round hash
+          auto generate_permutation = [&](auto round) {
+            auto &hash = std::get<round()>(current_hashes);
+            log_->debug("Using hash: {}", hash.toString());
+            auto &permutation = permutations_[round()];
+
+            std::seed_seq seed(hash.blob().begin(), hash.blob().end());
+            std::default_random_engine gen(seed);
+
+            permutation.resize(current_peers_.size());
+            std::iota(permutation.begin(), permutation.end(), 0);
+
+            std::shuffle(permutation.begin(), permutation.end(), gen);
+          };
+
+          generate_permutation(RoundTypeConstant<kCurrentRound>{});
+          generate_permutation(RoundTypeConstant<kNextRound>{});
+          generate_permutation(RoundTypeConstant<kRoundAfterNext>{});
+        }
+
+        auto peer = [this](auto round, auto pos) {
+          auto &permutation = permutations_[round];
+          // since reject round can be greater than number of peers, wrap it
+          // with number of peers
+          auto &peer = current_peers_[permutation[pos % permutation.size()]];
+          log_->debug("Using peer: {}", peer->toString());
+          return peer;
+        };
+
+        return {{peer(kCurrentRound, current_reject_round_ + 2),
+                 peer(kNextRound, 2),
+                 peer(kRoundAfterNext, 1),
+                 peer(kCurrentRound, current_reject_round_)}};
+      };
+
+      auto peers = notifier.get_observable()
+                       .with_latest_from(latest_hashes)
+                       .map(map_peers);
+
+      return std::make_shared<ordering::OnDemandConnectionManager>(
+          createNotificationFactory(std::move(async_call), delay), peers);
+    }
+
+    auto OnDemandOrderingInit::createGate(
+        std::shared_ptr<ordering::OnDemandOrderingService> ordering_service,
+        std::shared_ptr<ordering::transport::OdOsNotification> network_client,
+        std::unique_ptr<shared_model::interface::UnsafeProposalFactory>
+            factory) {
+      return std::make_shared<ordering::OnDemandOrderingGate>(
+          std::move(ordering_service),
+          std::move(network_client),
+          notifier.get_observable().map(
+              [](auto commit)
+                  -> ordering::OnDemandOrderingGate::BlockRoundEventType {
+                auto obs = commit.synced_blocks.as_blocking();
+                // if no blocks were commited
+                if (obs.count() == 0) {
+                  return ordering::OnDemandOrderingGate::EmptyEvent{};
+                } else {
+                  return obs.last();
```

Comment:
why we return only the last block to the notifier? We can miss some blocks information that were before the last block

Reply:
This information was not required before caching, so it can be refactored later.